Add table partitions to BigQuery

jeremywadsack opened the pull request:
See #1591 

This is an extension of @zedalaye's work to add the requested fixes.

I can squash and rebase after review.

Question: what do you think about using, e.g., `partition_type` instead of `time_partitioning_type` as the property name. That would be easier to read and understand (and matches the docs), although it diverges from the JSON API. I don't know which is more important to the gem.

Also, as @quartzmo mentioned, need a decision about whether to change expiration to seconds, as the rest of the gem uses, or retain milliseconds as is done with [`Dataset#default_expiration`](https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/9306c3840caf5bf106fdbd14249fffbc552effcd/google-cloud-bigquery/lib/google/cloud/bigquery/dataset.rb#L149-L172).

This should resolve #1627 as well.

quartzmo commented:
> what do you think about using, e.g., partition_type instead of time_partitioning_type as the property name. That would be easier to read and understand (and matches the docs), although it diverges from the JSON API. I don't know which is more important to the gem.

I am not opposed to the shorter name, but there is a case to be made for continuity with the official docs. Where do the docs use "partition type"?

quartzmo commented:
@jeremywadsack I don't know if you've signed a CLA, but be warned that googlebot has had trouble approving PRs with multiple contributors in the past, even when both have signed the CLA.

jeremywadsack commented:
> Where do the docs use "partition type"?

Oh, I just meant that the [general documentation](https://cloud.google.com/bigquery/docs/partitioned-tables) refers to "partitioned tables", never "time partitioning" as a concept. But I think all the api is `timePartitioning.type`.

jeremywadsack commented:
I did sign the CLA. As I'm adding to @zedalaye's work, do you have another approach? Can you accept and merge his work against `bigquery` branch and then I can add my changes separately?

jeremywadsack commented on google-cloud-bigquery/lib/google/cloud/bigquery/table.rb:
```diff
@@ -1217,6 +1239,58 @@ def to_gapi
+          #     table.time_partitioning_type = "DAY"
+          #   end
+          #
+          # @!group Schema
```
I don't really know what groups are in the docs, but this doesn't seem right as it's not really Schema. Should this be "Attributes" again here as well?

quartzmo replied:
Yes, please use `@!group Attributes` here.

quartzmo commented:
Since @zedalaye did not respond to my [comment](https://github.com/GoogleCloudPlatform/google-cloud-ruby/pull/1591#pullrequestreview-51540395) on #1591, my decision was to close the PR, and to soon begin prepare my own, completely new PR for this feature.

quartzmo commented:
> Can you accept and merge his work against bigquery branch and then I can add my changes separately?

PR #1591 is against master, and is not currently in an acceptable state.

quartzmo commented:
Sorry for the situation with googlebot! Unfortunately there is nothing we can do to help except to let you know what we've observed in the past.

jeremywadsack commented:
@googlebot I signed it!

quartzmo commented on google-cloud-bigquery/lib/google/cloud/bigquery/table.rb:
```diff
@@ -1217,6 +1239,58 @@ def to_gapi
+          # @!group Attributes
+          #
+          def time_partitioning_type= type
+            @gapi.time_partitioning ||= {}
```
I believe you can use `||= Google::Apis::BigqueryV2::TimePartitioning.new` here, and then set its `type` property.

quartzmo commented on google-cloud-bigquery/lib/google/cloud/bigquery/table.rb:
```diff
@@ -1217,6 +1239,58 @@ def to_gapi
+          # @!group Attributes
+          #
+          def time_partitioning_expiration= expiration
+            @gapi.time_partitioning ||= {}
```
I believe you can use `||= Google::Apis::BigqueryV2::TimePartitioning.new` here, and then set its `expiration_ms ` property.

jeremywadsack commented:
@quartzmo Thanks for being so responsive on this.

I pushed and update for those comments. Let me know if you see anything else.

I don't know what's going on with the CLA. https://cla.developers.google.com/clas says that there are none on file for me, but I completed one (on behalf of our company) and when I try to complete it again, it says "Looks like you've already signed this CLA".

quartzmo commented:
@jeremywadsack Regarding the CLA, feel free to try opening a brand new PR with just your commits, it may help.

blowmage commented:
Instead of opening a new PR, if you squash all commits on your branch and be the sole author then it can be approved.

blowmage commented on google-cloud-bigquery/lib/google/cloud/bigquery/table.rb:
```diff
@@ -118,6 +118,28 @@ def table_ref
+          time_partitioning = @gapi.time_partitioning
+          time_partitioning.expiration_ms if time_partitioning
+        end
+
```
I'd like to see a method with the following:

```ruby
def time_partitioning?
  !@gapi.time_partitioning.nil?
end
```

jeremywadsack replied:
Good thought. Done in bacb5ff.

jeremywadsack commented:
@blowmage I squashed commits and changed authorship in 53999c3.

I think we're just waiting on Google [to approve the CLA](https://cla.developers.google.com/about). 

> After the agreement is electronically signed, it will need to be reviewed by someone at Google. ... This normally takes a few days.

quartzmo commented:
@jeremywadsack Can you squash your commits into one? A good commit message would be something like:

```
Add time_partitioning attributes to Table

[closes #1627]
```

quartzmo commented on google-cloud-bigquery/acceptance/bigquery/table_test.rb:
```diff
@@ -70,6 +82,8 @@
     fresh.table?.must_equal true
     fresh.view?.must_equal false
+    fresh.time_partitioning_type.must_be_nil
+    fresh.time_partitioning_expiration_ms.must_be_nil
```
I get:

```sh
  2) Error:
Google::Cloud::Bigquery::Table::bigquery#test_0001_has the attributes of a table:
NoMethodError: undefined method `time_partitioning_expiration_ms' for #<Google::Cloud::Bigquery::Table:0x007f8299545ab0>
Did you mean?  time_partitioning_expiration
    google-cloud-bigquery/acceptance/bigquery/table_test.rb:86:in `block (2 levels) in <top (required)>'
```

quartzmo commented on google-cloud-bigquery/lib/google/cloud/bigquery/table.rb:
```diff
@@ -118,6 +118,35 @@ def table_ref
+        #
+        def time_partitioning_type
+          ensure_full_data!
+          time_partitioning.type if time_partitioning?
```
`time_partitioning.type` needs to be changed to `@gapi.time_partitioning.type`.

I get:

```sh
  3) Error:
Google::Cloud::Bigquery::Table::bigquery#test_0004_allow tables to be created with time_partioning enabled:
NameError: undefined local variable or method `time_partitioning' for #<Google::Cloud::Bigquery::Table:0x007f8299e6da20>
Did you mean?  time_partitioning?
               time_partitioning_type
    google-cloud-bigquery/lib/google/cloud/bigquery/table.rb:137:in `time_partitioning_type'
    google-cloud-bigquery/acceptance/bigquery/table_test.rb:139:in `block (2 levels) in <top (required)>'

quartzmo commented on google-cloud-bigquery/lib/google/cloud/bigquery/table.rb:
```diff
@@ -118,6 +118,35 @@ def table_ref
+        #
+        def time_partitioning_expiration
+          ensure_full_data!
+          time_partitioning.expiration_ms if time_partitioning?
```
`time_partitioning. expiration_ms ` needs to be changed to `@gapi.time_partitioning. expiration_ms `.

blowmage commented on google-cloud-bigquery/lib/google/cloud/bigquery/table.rb:
```diff
@@ -118,6 +118,90 @@ def table_ref
+
+
+        ###
+        # The expiration for the table partitions, if any,  in milliseconds.
```
This should be in seconds. Whenever we expose duration we do so in seconds.

jeremywadsack replied:
@blowmage Yes, well, almost always. :) See my comments in the PR about `Dataset#default_expiration` and @quartzmo's comments in #1591.

> Hmmm, although we generally present seconds, it looks like Dataset#default_expiration preserves the milliseconds of the underlying API. I suppose we will have to debate this in your PR.

> I'd say use milliseconds for now, and if there is a consensus that everything should be in seconds, then we'll have to discuss changing Dataset#default_expiration as well.

So let's open the debate. :)

quartzmo replied:
@blowmage Any idea why `Dataset#default_expiration` returns milliseconds?

blowmage replied:
Most likely an oversight.

quartzmo replied:
Can we change it? In that case I am OK with using seconds for the expiration attribute in this PR.

jeremywadsack replied:
Ok, I'll make that change.

quartzmo replied:
Sorry, I meant to ask @blowmage if can we change `Dataset#default_expiration` to return seconds.

quartzmo commented:
>  you "Cannot convert non-partitioned table to partitioned table." So I'm not sure if these should have writer attributes on table

Hmmm, what about updating the expiration to a different value on a partitioned table? If that works, my opinion is that we're good, and can rely on the service error you got to inform people. If they ever add another supported value for `type`, we can add that to the update test later.

quartzmo commented on google-cloud-bigquery/lib/google/cloud/bigquery/table.rb:
```diff
@@ -141,6 +141,10 @@ def time_partitioning_type
         # ](https://cloud.google.com/bigquery/docs/partitioned-tables).
         #
+        # You can only set partitioning when creating a table as in
+        # the example below. BigQuery does not allow you to change partiting
```
typo: "partiting"

But since you have to change it anyway, what about adapting the error message?

"BigQuery does not allow you to convert a non-partitioned table to a partitioned table."

jeremywadsack replied:
I really seem to have trouble typing that word. I'll fix it. :/

I went with this statement rather than what you suggest (which was my first pass) because it also does not allow you to change a partitioned table to an unpartitioned one. In that case the error message is:

> Google::Cloud::InvalidArgumentError: invalid: For time based partitioning, only the partition expiration time can be changed.

But I can still change the statement if that's what you want to say.

quartzmo replied:
No, your reasoning is good, just fix the spelling!

quartzmo commented:
> Rubocop thinks this is easier to read?!

Always good times with Rubocop! 😆

quartzmo commented on google-cloud-bigquery/lib/google/cloud/bigquery/table.rb:
```diff
@@ -118,6 +118,96 @@ def table_ref
+        # ](https://cloud.google.com/bigquery/docs/partitioned-tables).
+        #
+        # You can only set partitioning when creating a table as in
+        # the example below. BigQuery does not allow you to change partiting
```
s/partiting/partitioning/

jeremywadsack commented:
@quartzmo Fixed spelling, rebased, and squashed.

@blowmage CLA approved this morning.

quartzmo commented:
Rebuilding CI (failure in pubsub, see #1653).

quartzmo commented:
@jeremywadsack I'm not sure if you have appetite for more, but I looked through the upcoming BigQuery additions and found [a few more straightforward features](https://github.com/GoogleCloudPlatform/google-cloud-ruby/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+bug%22), if you're interested.

jeremywadsack commented:
Thank for being so responsive and helping me get this done.

I'm not sure that those are features that we need and I've just started to use BigQuery. If these are intrinsic to our project, I'll take a look.

quartzmo commented:
@jeremywadsack Sounds good. Please keep in touch and let us know (open an issue) if you need anything.
